lib/execution-info.js

```diff
@@ -0,0 +1,526 @@
+const errors = require('./errors');
+
+/**
+ * A base class that represents a wrapper around the user provided query options with getter methods and proper
+ * default values.
+ */
+class ExecutionInfo {
+
+  /**
+   * Creates a new instance of {@link ExecutionInfo}.
+   */
+  constructor() {
+  }
+
+  /**
+   * Creates an empty instance, where all methods return undefined, used internally.
+   * @ignore
+   * @return {ExecutionInfo}
+   */
+  static empty() {
+    return new ExecutionInfo();
+  }
+
+  /**
+   * Determines if the stack trace before the query execution should be maintained.
+   * @abstract
+   * @returns {Boolean}
+   */
+  getCaptureStackTrace() {
+
+  }
+
+  /**
+   * Gets the [Consistency level]{@link module:types~consistencies} to be used for the execution.
+   * @abstract
+   * @returns {Number}
+   */
+  getConsistency() {
+
+  }
+
+  /**
+   * Key-value payload to be passed to the server. On the server side, implementations of QueryHandler can use
+   * this data.
+   * @abstract
+   * @returns {Object}
+   */
+  getCustomPayload() {
+
+  }
+
+  /**
+   * Gets the amount of rows to retrieve per page.
+   * @abstract
+   * @returns {Number}
+   */
+  getFetchSize() {
+
+  }
+
+  /**
+   * When a fixed host is set on the query options and the query plan for the load-balancing policy is not used, it
+   * gets the host that should handle the query.
+   * @returns {Host}
+   */
+  getFixedHost() {
+
+  }
+
+  /**
+   * Gets the type hints for parameters given in the query, ordered as for the parameters.
+   * @abstract
+   * @returns {Array|Array<Array>}
+   */
+  getHints() {
+
+  }
+
+  /**
+   * Determines whether the driver must retrieve the following result pages automatically.
+   * <p>
+   *   This setting is only considered by the [Client#eachRow()]{@link Client#eachRow} method.
+   * </p>
+   * @abstract
+   * @returns {Boolean}
+   */
+  getIsAutoPage() {
+
+  }
+
+  /**
+   * Determines whether its a counter batch. Only valid for [Client#batch()]{@link Client#batch}, it will be ignored by
+   * other methods.
+   * @abstract
+   * @returns {Boolean}
+   */
+  getIsBatchCounter() {
+
+  }
+
+  /**
+   * Determines whether the batch should be written to the batchlog. Only valid for
+   * [Client#batch()]{@link Client#batch}, it will be ignored by other methods.
+   * @abstract
+   * @returns {Boolean}
+   */
+  getIsBatchLogged() {
+
+  }
+
+  /**
+   * Determines whether the query can be applied multiple times without changing the result beyond the initial
+   * application.
+   * @abstract
+   * @returns {Boolean}
+   */
+  getIsIdempotent() {
+
+  }
+
+  /**
+   * Determines whether the query must be prepared beforehand.
+   * @abstract
+   * @return {Boolean}
+   */
+  getIsPrepared() {
+
+  }
+
+  /**
+   * Determines whether query tracing is enabled for the execution.
+   * @abstract
+   * @returns {Boolean}
+   */
+  getIsQueryTracing() {
+
+  }
+
+  /**
+   * Gets the keyspace for the query.
```

Comment:
For properties that might be null like this one and paging state, should we also indicate that they may return null if unset?

Reply:
I've included a note at class level (possible `undefined` return values) and some info regarding what it means for keyspace to be `undefined`.

Most values depend on the default values, that can be defined by the user when creating the `Client` instance,  so I find it hard to document possible return values... :)

TLDR return values can be `undefined` with few exceptions (lbp, retry policy, some flags, ...)